core/src/main/kotlin/ch/ergon/dope/resolvable/expression/CaseExpression.kt

```diff
@@ -0,0 +1,132 @@
+private const val ELSE = "ELSE"
+
+sealed interface CaseExpression : Expression
+
+sealed interface UnaliasedCaseExpression : CaseExpression {
+    fun alias(alias: String) = AliasedCaseExpression(alias, this)
+}
+
+class AliasedCaseExpression(
+    private val alias: String,
+    private val unaliasedCaseExpression: UnaliasedCaseExpression,
+) : CaseExpression {
+    override fun toDopeQuery(): DopeQuery {
+        val caseDopeQuery = unaliasedCaseExpression.toDopeQuery()
+
+        val queryString = caseDopeQuery.queryString.trimEnd()
+            .removeSuffix(END)
+            .plus("AS `$alias` $END")
+
+        return DopeQuery(queryString, caseDopeQuery.parameters)
+    }
+}
+
+class CaseClass<T : ValidType>(val case: UnaliasedExpression<T>) : Resolvable {
+    override fun toDopeQuery(): DopeQuery {
+        val caseDopeQuery = case.toDopeQuery()
+        return DopeQuery("$CASE ${caseDopeQuery.queryString} $END", caseDopeQuery.parameters)
+    }
+}
+
+class SimpleCaseExpression<T : ValidType>(
+    val case: UnaliasedExpression<T>,
+    val firstWhenThen: Pair<TypeExpression<T>, UnaliasedExpression<out ValidType>>,
+    val whenThenMap: Map<TypeExpression<T>, UnaliasedExpression<out ValidType>> = emptyMap(),
+) : UnaliasedCaseExpression {
+    override fun toDopeQuery(): DopeQuery {
+        val caseDopeQuery = case.toDopeQuery()
+        val dopeQueries = (mapOf(firstWhenThen) + whenThenMap).map { it.key.toDopeQuery() to it.value.toDopeQuery() }
+        return DopeQuery(
+            queryString = "$CASE ${caseDopeQuery.queryString} ${dopeQueries.joinToString(" ")
+            { "$WHEN ${it.first.queryString} $THEN ${it.second.queryString}" }
+            } $END",
+            parameters = emptyMap(),
+        )
+    }
+}
+
+class SimpleElseCaseExpression<T : ValidType>(
+    private val case: UnaliasedExpression<T>,
+    private val firstWhenThen: Pair<TypeExpression<T>, UnaliasedExpression<out ValidType>>,
+    private val whenThenMap: Map<TypeExpression<T>, UnaliasedExpression<out ValidType>> = emptyMap(),
+    private val elseCase: UnaliasedExpression<out ValidType>,
+) : UnaliasedCaseExpression {
+    override fun toDopeQuery(): DopeQuery {
+        val caseDopeQuery = case.toDopeQuery()
+        val dopeQueries = (mapOf(firstWhenThen) + whenThenMap).map { it.key.toDopeQuery() to it.value.toDopeQuery() }
+        val elseCaseDopeQuery = elseCase.toDopeQuery()
+        return DopeQuery(
+            queryString = "$CASE ${caseDopeQuery.queryString} ${dopeQueries.joinToString(" ")
+            { "$WHEN ${it.first.queryString} $THEN ${it.second.queryString}" }
+            }${elseCaseDopeQuery.let { " $ELSE ${elseCaseDopeQuery.queryString}" }
+            } $END",
+            parameters = emptyMap(),
+        )
+    }
+}
+
+fun <T : ValidType> case(case: UnaliasedExpression<T>) = CaseClass(case)
+
+fun <T : ValidType> CaseClass<T>.`when`(whenCondition: TypeExpression<T>, thenExpression: UnaliasedExpression<out ValidType>) =
```

Comment:
I do not like that the extension functions have to be used with backticks. We could use other keywords, I know that means we won't be exactly like jOOQ 🙂

Reply:
Do you have a suggestion?

Reply:
The following naming will be implemented:
when -> whenThen
else -> otherwise